my/core/query.py

```diff
@@ -0,0 +1,717 @@
+
+
+def test_parse_timedelta_string():
+
+    import pytest
+
+    with pytest.raises(ValueError) as v:
+        parse_timedelta_string("5xxx")
+
+    assert v is not None
+    assert str(v.value).startswith("Could not parse time duration from")
+
+    res = parse_timedelta_string("1w5d5h10m50s")
+    assert res == timedelta(days=7.0 + 5.0, hours=5.0, minutes=10.0, seconds=50.0)
+
+
+# classes to use in tests, need to be defined at the top level
+# because of a mypy bug
+class _Int(NamedTuple):
+    x: int
+
+
+# to test order_key with compatible orderable (int, float) types
+class _Float(NamedTuple):
+    x: float
+
+
+def test_basic_orders() -> None:
+
+    import random
+
+    import pytest
+
+    def basic_iter() -> Iterator[_Int]:
+        for v in range(1, 6):
+            yield _Int(v)
+
+    def filter_two(obj: Any) -> bool:
+        return obj.x != 2
+
+    res = list(select(basic_iter(), where=filter_two, reverse=True))
+    assert len(res) == 4
+
+    for (actual, expected) in zip(res, (5, 4, 3, 1)):
+        assert actual == _Int(expected)
+
+
+    input_items = list(basic_iter())
+    random.shuffle(input_items)
+
+    res = list(select(input_items, order_key="x"))
+    assert len(res) == 5
+
+    for (actual, expected) in zip(res, (1, 2, 3, 4, 5)):
+        assert actual.x == expected  # type: ignore
+
+    # default int ordering
+    def custom_order_by(obj: Any) -> Any:
+        return getattr(obj, "x")
+
+    # sort random ordered list, only return first two items
+    res = list(select(input_items, where=filter_two, order_by=custom_order_by, limit=2))
+    assert len(res) == 2
+
+    for (actual, expected) in zip(res, (1, 3)):
+        assert actual == _Int(expected)
+
+    # filter produces empty iterator
+    with pytest.warns(UserWarning, match=r"encountered empty iterable"):
+        res = list(select(input_items, where=lambda o: o is None, order_key="x"))
+    assert len(res) == 0
+
+
+def test_order_key_multi_type() -> None:
+
+    def basic_iter() -> Iterator[_Int]:
+        for v in range(1, 6):
+            yield _Int(v)
+
+
+    def floaty_iter() -> Iterator[_Float]:
+        for v in range(1, 6):
+            yield _Float(float(v + 0.5))
+
+
+    res = list(select(itertools.chain(basic_iter(), floaty_iter()), order_key="x"))
+    assert len(res) == 10
+
+    for (actual, expected) in zip(res, (1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 5.5)):
+        assert actual.x == expected  # type: ignore
+
+
+def test_couldnt_determine_order() -> None:
+
+    res = list(select(iter([object()]), order_value=lambda o: isinstance(o, datetime)))
+    assert len(res) == 1
+    assert isinstance(res[0], Unsortable)
+    assert type(res[0].obj) == object
+
+
+# same value type, different keys, with clashing keys
+class _A(NamedTuple):
+    x: datetime
+    y: int
+    z: int
+
+
+class _B(NamedTuple):
+    y: datetime
+
+
+# move these to tests/? They are re-used so much in the tests below,
+# not sure where the best place for these is
```

Comment:
yeah -- I guess it's nice to have some basic test[s] near the code, as a sort of demonstration, but maybe the details should go in the actual test file? Although still looks OK to me, maybe with more core tests it will become clearer what's the best way